# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [3]:
os.getcwd()

'C:\\Users\\Suraj Kumar\\Desktop\\Python\\Guided Capstone\\Step 2'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [35]:
df=pd.read_csv('step2_output.csv')
df.drop(['Unnamed: 0'],axis=1,inplace=True)
new_df=df

In [36]:
df.head()

Name   Region    state  summit_elev  vertical_drop  \
0       Alyeska Resort   Alaska   Alaska         3939           2500   
1  Eaglecrest Ski Area   Alaska   Alaska         2600           1540   
2     Hilltop Ski Area   Alaska   Alaska         2090            294   
3     Arizona Snowbowl  Arizona  Arizona        11500           2300   
4  Sunrise Park Resort  Arizona  Arizona        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  LongestRun_mi  \
0        250      1        0.0          0          2  ...            1.0   
1       1200      0        0.0          0          0  ...            2.0   
2       1796      0        0.0          0          0  ...            1.0   
3       9200      0        0.0          1          0  ...            2.0   
4       9200      0        0.0          0          1  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [37]:
df1=df[['state']]

In [38]:
df = pd.concat([df.drop(df1, axis=1), pd.get_dummies(df1)], axis=1)

In [39]:
df.drop('Region',axis=1,inplace=True)

In [40]:
df

Name  summit_elev  vertical_drop  base_elev  \
0                       Alyeska Resort         3939           2500        250   
1                  Eaglecrest Ski Area         2600           1540       1200   
2                     Hilltop Ski Area         2090            294       1796   
3                     Arizona Snowbowl        11500           2300       9200   
4                  Sunrise Park Resort        11100           1800       9200   
..                                 ...          ...            ...        ...   
325               Meadowlark Ski Lodge         9500           1000       8500   
326          Sleeping Giant Ski Resort         7428            810       6619   
327                   Snow King Resort         7808           1571       6237   
328  Snowy Range Ski & Recreation Area         9663            990       8798   
329                White Pine Ski Area         9500           1100       8400   

     trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0        1        0.0          0          2     2       0  ...   
1        0        0.0          0          0     0       0  ...   
2        0        0.0          0          0     0       1  ...   
3        0        0.0          1          0     2       2  ...   
4        0        0.0          0          1     2       3  ...   
..     ...        ...        ...        ...   ...     ...  ...   
325      0        0.0          0          0     0       1  ...   
326      0        0.0          0          0     0       1  ...   
327      0        0.0          0          0     1       1  ...   
328      0        0.0          0          0     0       1  ...   
329      0        0.0          0          0     0       2  ...   

     state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
0                     0                   0                0           0   
1                     0                   0                0           0   
2                     0                   0                0           0   
3                     0                   0                0           0   
4                     0                   0                0           0   
..                  ...                 ...              ...         ...   
325                   0                   0                0           0   
326                   0                   0                0           0   
327                   0                   0                0           0   
328                   0                   0                0           0   
329                   0                   0                0           0   

     state_Vermont  state_Virginia  state_Washington  state_West Virginia  \
0                0               0                 0                    0   
1                0               0                 0                    0   
2                0               0                 0                    0   
3                0               0                 0                    0   
4                0               0                 0                    0   
..             ...             ...               ...                  ...   
325              0               0                 0                    0   
326              0               0                 0                    0   
327              0               0                 0                    0   
328              0               0                 0                    0   
329              0               0                 0                    0   

     state_Wisconsin  state_Wyoming  
0                  0              0  
1                  0              0  
2                  0              0  
3                  0              0  
4                  0              0  
..               ...            ...  
325                0              1  
326                0              1  
327                0              1  
328                0              1  
329         

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we don't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [11]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing
#from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [12]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [13]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [14]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [15]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [16]:
explained_variance_score(y_test,y_pred)

0.925555553486166

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [17]:
print(lm.intercept_)

64.11118726323181


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [19]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending= False).head(15)


Coefficient
total_chairs         8.904666e+13
fastQuads            3.375779e+13
surface              3.162852e+13
double               2.787222e+13
triple               2.486394e+13
state_New York       2.241803e+13
state_Michigan       2.115653e+13
quad                 2.015130e+13
state_Colorado       1.864013e+13
state_California     1.824110e+13
state_Pennsylvania   1.740681e+13
state_New Hampshire  1.605044e+13
state_Wisconsin      1.605044e+13
state_Vermont        1.556550e+13
state_Minnesota      1.506155e+13

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [45]:
ndf.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'base_elev', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac'],
      dtype='object')

In [42]:
ndf=new_df.drop(['Region','state'],axis=1)

In [47]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X1 = ndf.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y1 = ndf.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler1 = preprocessing.StandardScaler().fit(X1)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled1=scaler1.transform(X1)

In [48]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y1 = y1.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
x1_train, x1_test, y1_train, y1_test = train_test_split(X_scaled1, y1, test_size=0.25, random_state=1)

In [50]:
nlm = linear_model.LinearRegression()
model1 = nlm.fit(x1_train,y1_train)
pred1 = model1.predict(x1_test)

In [52]:
explained_variance_score(y1_test, pred1)

0.9244193604833097

In [53]:
mean_absolute_error(y1_test, pred1)

5.5099192670842605

In [54]:
nlm.intercept_

64.03936971345686

In [56]:
pd.DataFrame(abs(nlm.coef_), X1.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending= False).head(10)

Coefficient
AdultWeekday        20.224262
summit_elev         12.543815
base_elev            9.736447
vertical_drop        4.351901
averageSnowfall      1.929602
quad                 1.552442
triple               1.388252
surface              1.254160
daysOpenLastYear     1.184864
Runs                 1.019630

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [57]:
ndf=new_df.drop(['Region','state','summit_elev','base_elev'],axis=1)

In [58]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X1 = ndf.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y1 = ndf.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler1 = preprocessing.StandardScaler().fit(X1)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled1=scaler1.transform(X1)

In [59]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y1 = y1.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
x1_train, x1_test, y1_train, y1_test = train_test_split(X_scaled1, y1, test_size=0.25, random_state=1)

In [60]:
nlm = linear_model.LinearRegression()
model1 = nlm.fit(x1_train,y1_train)
pred1 = model1.predict(x1_test)

In [61]:
explained_variance_score(y1_test, pred1)

0.9257043387913915

In [63]:
mean_absolute_error(y1_test, pred1)

5.477922321886279

In [64]:
nlm.intercept_

64.04022316513387

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.925555553486166 |  |-|
| Model 2. | 0.9244193604833097| |'state'|
| Model 3. | 0.9257043387913915 | |'state','summit_elev','base_elev'|

Model Selection:

In [ ]:
Model 3 will selected